### PR TITLE
feat(linter) eslint-plugin-import: no-amd

### PR DESCRIPTION
**Pull request by qinyuhang**
Hi, this PR implements https://github.com/import-js/eslint-plugin-import/blob/main/src/rules/no-amd.js, And this is my first time of contribute to OXC repo. PLZ comment anything that I am doing not correct. THX.

### CODE REVIEW

File: crates/oxc_linter/src/rules.rs [hunk 1]
```diff
@@ -311,4 +312,6 @@ oxc_macros::declare_all_lint_rules! {
     import::no_cycle,
     import::no_self_import,
     jsx_a11y::alt_text
+    import::no_amd,
+    jsx_a11y::alt_text,
```
Comment (camc314) [inline crates/oxc_linter/src/rules.rs]: don't think we need the duplicate `jsx_a11y::alt_text`?

File: crates/oxc_linter/src/rules/import/no_amd.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::{Atom, ModuleKind, Span};
+use oxc_ast::AstKind;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint-plugin-import(no-amd): Do not use AMD `require` and `define` calls.")]
+#[diagnostic(severity(warning), help("Consider change {1} to EMScript import or export"))]
```
Comment (camc314) [inline crates/oxc_linter/src/rules/import/no_amd.rs:14]:
```suggestion
#[diagnostic(severity(warning), help("Consider changing {1} to ECMAScript import or export"))]

File: crates/oxc_linter/src/rules/import/no_amd.rs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    ];
+
+    let fail = vec![
+      "require([a, b], function() {})",
+      "define([a, b], function() {})",
+    ];
+
+    tester = tester.change_rule_path("no-amd.js").update_expect_pass_fail(pass, fail);
+    tester.test_and_snapshot();
+  }
```
Comment (camc314) [inline crates/oxc_linter/src/rules/import/no_amd.rs:92]: just do like this?
```rs
    let pass = vec![
      "var _ = require('lodash')",
      "var find = require('lodash.find')",
      "var foo = require('./foo')",
      "var foo = require('../foo')",
      "var foo = require('foo')",
      "var foo = require('./')",
      "var foo = require('@scope/foo')",
      "var bar = require('./bar/index')",
    ];

    let fail = vec![
      "require([a, b], function() {})",
      "define([a, b], function() {})",
    ];

Tester::new_without_config::<String>(NoAmd::NAME, pass, fail)
    .with_import_plugin(true).test_and_snapshot

**Review by camc314 — COMMENTED** [review]
Looks great! - Thanks for contributing

File: crates/oxc_linter/src/rules/import/no_amd.rs (hunk 3 above)
Comment (qinyuhang) [reply crates/oxc_linter/src/rules/import/no_amd.rs]: Thanks, I updated the codes like you suggested.

File: crates/oxc_linter/src/rules/import/no_amd.rs (hunk 2 above)
Comment (qinyuhang) [reply crates/oxc_linter/src/rules/import/no_amd.rs]: Thanks, I updated the codes like you suggested.

File: crates/oxc_linter/src/rules/import/no_amd.rs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    let call_expr = match node.kind() {
+      AstKind::CallExpression(call_expr) => call_expr,
+      _ => return,
+    };
+
+    let identifier = match &call_expr.callee {
+      Expression::Identifier(ref xx) => xx,
+      _ => return,
+    };
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/import/no_amd.rs:54]: All these 3 matches could be written with a simple if let else.

File: crates/oxc_linter/src/rules/import/no_amd.rs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    "var foo = require('foo')",
+    "var foo = require('./')",
+    "var foo = require('@scope/foo')",
+    "var bar = require('./bar/index')",
+  ];
+
+  let fail = vec![
+    "require([a, b], function() {})",
+    "define([a, b], function() {})",
+  ];
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/import/no_amd.rs:83]: Shall we copy all the tests over from https://github.com/import-js/eslint-plugin-import/blob/main/tests/src/rules/no-amd.js ?

**Review by Boshen — CHANGES_REQUESTED** [review]

File: crates/oxc_linter/src/rules/import/no_amd.rs (hunk 5 above)
Comment (qinyuhang) [reply crates/oxc_linter/src/rules/import/no_amd.rs:83]: OK, I'll update it later. Thanks.

File: crates/oxc_linter/src/rules/import/no_amd.rs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    /// require(`../name`);
+    /// ```
+    NoAmd,
+    nursery
+);
+
+/// https://github.com/import-js/eslint-plugin-import/blob/main/src/rules/no-amd.js
+impl Rule for NoAmd {
+  fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+    let _ = match ctx.source_type().module_kind() {
```
Comment (qinyuhang) [inline crates/oxc_linter/src/rules/import/no_amd.rs:41]: It seems I have a misunderstanding of `ctx.source_type().module_kind()`. The original code was `if (context.getScope().type !== 'module') { return; }`. Does OXC has any function similar to the original one? Current code is incorrect and will cause test case `function x() { define(["a"], function (a) {}) }` fail, which should pass.

Comment (qinyuhang) [reply crates/oxc_linter/src/rules/import/no_amd.rs]: Try compare the `ctx.scopes().root_scope_id()` and `node.scope_id()`. if they are the same, this means the CallExpression is in the top level.

File: crates/oxc_linter/src/rules/import/no_amd.rs (hunk 2 above)
Comment (qinyuhang) [reply crates/oxc_linter/src/rules/import/no_amd.rs]: Updated.

File: crates/oxc_linter/src/rules/import/no_amd.rs (hunk 5 above)
Comment (qinyuhang) [reply crates/oxc_linter/src/rules/import/no_amd.rs:105]: Updated by copy all tests here.

**Review by Boshen — APPROVED** [review]

**Pull request merged**